language/move-package/src/package.rs

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) The Diem Core Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::Result;
+use std::collections::BTreeMap;
+
+use bytecode_verifier::{cyclic_dependencies, dependencies, verify_module};
+use diem_types::vm_status::StatusCode;
+use move_core_types::language_storage::ModuleId;
+use vm::{
+    access::ModuleAccess,
+    errors::{BinaryLoaderResult, Location, PartialVMError, VMResult},
+    file_format::CompiledModule,
+};
+
+pub struct CompiledPackage {
+    name: String,
+    modules_with_bytes: Vec<(Vec<u8>, CompiledModule)>,
+}
+
+#[derive(Clone)]
+pub struct VerifiedPackage {
+    name: String,
+    modules_with_bytes: Vec<(Vec<u8>, CompiledModule)>,
+    dependencies: Vec<VerifiedPackage>,
+}
```

Comment:
Thanks for the comments Victor!

> packages could form a dependency graph that is not a tree.

I tend to think this is not the case (although in theory it could be), and the reason why I tend to think that packages also form a tree instead of a graph is based on the property that there should be no cyclic dependencies among the modules, and hence, if the packages are grouped reasonably, there should be have no cyclic dependencies among the packages too.

> You want the client to be able to use A with other packages, potentially ones you don't control, then there is no way you can be sure the client uses the version of B owned by A.

I totally agree with you. I have some thoughts on this and the arguments pretty much depend on how the modules are deployed (or rather, published).

- Given the current model, the modules on-chain is the source of truth, and there is no guarantee that what you put in the `dependencies` is the same as the modules on-chain. In that sense, yes, it is a false sense of verified. However, I would argue that the property of having the same on-chain modules it not the problem addressed by a `VerifiedPackage`, it should be enforced by whomever has the access of on-chain modules, and currently, the VM.
  - As a further extension, if we were to have package publishing (e.g., multiple modules in one transaction), the VM could check the dependencies declared in a `VerifiedPackage` against all on-chain modules, and reject if
    - (strict mode) modules mismatch in any byte
    - (compatibility mode) modules that differ do not pass the compatibility check.
- Image there could be an off-chain use case (although that is not the use case I intended), then there is no source-of-truth of modules. Every module that you put in the `VerifiedPackage` is the source of truth and the package is guaranteed to be self-contained.

Reply:
> a tree instead of a graph is based on the property that there should be no cyclic dependencies among the modules
Sorry I didn't make this clear. By "not a tree" I meant something like a directed acyclic graph:
![graphviz (6)](https://user-images.githubusercontent.com/10379359/108924533-a91fd000-75ef-11eb-88e1-c3aa1f173092.png)

> the arguments pretty much depend on how the modules are deployed (or rather, published).
I could be wrong, but I feel like concepts like "on-chain" or "module publishing" are orthogonal in this context. Let me explain what I mean by this.

I wasn't really complaining about dependencies of a (verified) package not matching those on-chain. In fact I think it's a different issue and totally agree with you that it's not something the package system should worry about. I also don't see this as a matter of defining the one and only source of truth.

All I was saying was that I didn't feel like `VerifiedModule` was the model clients would want and it would likely create more problems than it would solve. IMO the biggest issue is that `VerifiedModule` is trying to prove an invariant that's broader than its own scope. 

Let's first clarify the property a typical client wants to uphold:
  - "for a given universe of modules, make sure they verify and link"

Here the property applies to a bigger universe, which could contain modules from various sources, for example, different packages or even modules created on the fly. Now trying to impose the property on a single package will result in the package carrying a bigger universe, or even make it the universe itself, which kinda defeats the purpose of having packages in the first place. Here's a metaphor: all I need is a wheel, but I have to pay for a full car just to demonstrate that the wheel works, even when I just want to install that wheel on another car I already have. 

In other words, imagine what a taking the verified package as parameters:
- If it's `&[VerifiedPackage]`, then how do you deal with `[A -> B, B', C -> B'']`, where `B, B', B''` are all compatible but semantically different, or even worse, where `B, B', B''` are in fact incompatible? In this case the per package verification is not only redundant but could create additional confusion.
- If you force clients to take only one `&VerifiedPackage`, then it's the universe, not a package.

Also consider the point @Tim made in the meeting: packaging is a build system concept, not a language one. I think its role is to provide an easier way to refer to or move things as a bundle and it shouldn't care about syntactic or semantic correctness -- that's the job of the client/consumer.